oddsnends/realtime/mbplotfilter.1
.TH mbplotfilter 1 "1 March 1994"
.SH NAME
mbplotfilter - Produce L-DGO metafile graphics from output 
of mbcontourfilter.

.SH VERSION
Version 4.00

.SH SYNOPSIS
\fBmbplotfilter\fP \fB-R\fIwest/east/south/north [-Wwidth -V]

.SH DESCRIPTION
\fBmbplotfilter\fP reads ascii plot calls from stdin and
outputs L-DGO metafile graphics to stdout. It is intended to
be used with mbcontourfilter to plot real-time Hydrosweep data
on board the R/V Maurice Ewing. See the manual pages for 
\fBmbcontourfilter\fP and \fBmbm_calcomp_plot\fP.

.SH AUTHORSHIP
David W. Caress (caress@lamont.ldeo.columbia.edu)
.br
Dale N. Chayes (dale@lamont.ldeo.columbia.edu)
.br
Lamont-Doherty Earth Observatory
.br
Palisades, NY 10964

.SH OPTIONS
.TP
.B \fB-R\fIwest/east/south/north\fP
Sets the longitude and latitude bounds of the plotting.  A mercator
projection is used.
.TP
.B \fB-V\fIverbose\fP
Causes status messages to be printed to stderr as the program runs.
.TP
.B \fB-W\fIwidth\fP
Sets the width in inches of the plot.  A mercator projection is used
to interprete the longitude and latitude values input in the
plot calls.

.SH EXAMPLES
On the R/V Maurice Ewing, the real-time hydrosweep data 
is logged in a file which changes each day.  In order to make a 
contour pen plot of the real time data in a file called
/net/moray/logger/data/9305hs.d237 the following approach can
be used:
 	tail -0f /net/moray/logger/data/9305hs.d237 | mbcontourfilter 
 		-F5 -R-43.0/-42.0/10.08333/10.916667 -J30 
 		-C50/200/200/200/.02/.07 -A0.25/1.0/4.0 | mbplotfilter 
 		 -R-43.0/-42.0/10.08333/10.916667 -W30 | ccgr -d /dev/ttyb
.br
As invoked, the tail command will read through the input file until it
reaches the end and then enter an infinite loop, reading new 
data as it is added to the end of the file.  The data is piped to 
\fBmbcontourfilter\fP, which generates contours, shiptracks, and 
annotations as ascii plot calls printed to stdout.  The ascii plot calls
are piped to \fBmbplotfilter\fP, which interprets them in terms of
Lamont metafile graphics calls.  The metafile graphics are piped to
a program called ccgr, which translates them to Calcomp pen
plotter commands and sends those commands to a serial port.  Assuming
that the plotter is connected to the port used, the plotter then
draws the contours on paper.

In order to use this system for pen plotting in other environments or
with other pen plotters, the user would need to replace mbplotfilter and
ccgr with one or more programs to translate the ascii plot calls to
the required plot call format.

The format of the plot calls understood by \fBmbplotfilter\fP is
simple.  Four calls are understood:
 	init
 	newp ipen
 	plot x y penup
 	stop
.br
Here ipen is the pen number to be used (usually 1-4), x and y are the plot
coordinates in longitude and latitude, and penup indicates whether
the pen is up (penup=3) or down (penup=2) during the pen move.

.SH BUGS
\fBmbplotfilter\fP works only with plot calls from \fBmbcontourfilter\fP.


NEWS
-------------------------------------------------------------------------------
MB-SYSTEM NEWS FILE:
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

In the GNU package standard, the NEWS file lists the history of user-visible
changes to a package. For the the MB-System open source software package for the
processing and display of swath sonar data, this information is included in the
ChangeLog file. Both the NEWS file and the ChangeLog file are located at the top
of the MB-System source code distribution directory structure.

-------------------------------------------------------------------------------


oddsnends/realtime/mbplotfilter.c
/*
 *	MBPLOTFILTER is a filter to take the output of mb_realtime and put
 *	it into Lamont graphics metafile format.  Essentially, it
 *	reads the plot calls made by mb_contfilter and in turn
 *	makes the appropriate calls to the Lamont graphics library.
 *
 *	David W. Caress
 *	L-DGO
 *	May 1991
 */
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <graphics.h>

#define IUP 3
#define IDN 2

char	line[200];
float	width, xmin,xmax,ymin,ymax;
float	homex = 25.;
float	homey = 25;
float	x,y;
int	verbose;
int	ipen;
int	fd;
int	status;
int	istart,iplotting;
float	cmerc, er, pr, scl;
float     mtx[3][2];

main(argc, argv)
int argc;
char **argv; 
{
	int	i;

	/* set defaults */
	width = 24.0;	/* in inches */
	xmin = 0.0;
	xmax = 0.0;
	ymin = 0.0;
	ymax = 0.0;
	verbose = 0;

	/* process argument list */
	for (i = 1; i < argc; i++) 
		{
		if (argv[i][0] == '-') 
			{
			switch (argv[i][1]) 
				{
				case 'R':
				case 'r':
					sscanf (&argv[i][2], "%f/%f/%f/%f", 
						&xmin,&xmax,&ymin,&ymax);
					break;
				case 'W':
				case 'w':
					sscanf (&argv[i][2], "%f",
						&width);
					break;
				case 'V':
				case 'v':
					verbose = 1;
					break;
				}
			}
		}

	/* check for bounds */
	if (xmin == xmax || ymin == ymax)
		{
		fprintf(stderr,"mb_plotfilter error: bounds not set or set incorrectly\n");
		exit(-1);
		}

	/* read until init call is made */
	istart = 1;
	while (istart)
		{
		fgets(line,sizeof(line),stdin);
		if (strncmp(line,"init",4) == 0)
			{
			istart = 0;
			initg();
			fd = 1;
			mfon(fd,0);
			cmerc = (xmin +xmax)/2.;
			scl = -width/(xmax-xmin);
			er = 0.;
			pr = 0.;
			fprintf(stderr,"calling vmerc: %f %f %f %f\n",
				cmerc,scl,er,pr);
			vmerc_(&cmerc,&scl,&er,&pr);
			fprintf(stderr,"done vmerc: %f %f %f %f\n",
				cmerc,scl,er,pr);
			x = xmin;
			y = ymin;
			fprintf(stderr,"calling xymerc: xmin:%f ymin:%f\n",x,y);
			xymerc_(&x,&y);
			fprintf(stderr,"done xymerc: xmin:%f ymin:%f\n:",x,y);
			fprintf(stderr,"calling initm2: [0][0]:%d [0][1]:%d %d [1][0]:%d [1][1]:%d %d [2][0]:%d [2][1]:%d %d\n",
				mtx[0][0],mtx[0][1],
				mtx[1][0],mtx[1][1],
				mtx[2][0],mtx[2][1]);
			initm2(mtx);
			fprintf(stderr,"done initm2: [0][0]:%d [0][1]:%d %d [1][0]:%d [1][1]:%d %d [2][0]:%d [2][1]:%d %d\n",
				mtx[0][0],mtx[0][1],
				mtx[1][0],mtx[1][1],
				mtx[2][0],mtx[2][1]);
			fprintf(stderr,"calling tlate2: %f %f\n",-x,-y);
			tlate2(mtx,-x,-y);
			fprintf(stderr,"done tlate2: %f %f\n",-x,-y);
			makcur();

			/* init message */
			if (verbose)
				{
				fprintf(stderr,"\nMBPLOTFILTER:\n");
				fprintf(stderr,"bounds: %f %f %f %f\n",xmin,xmax,ymin,ymax);
				fprintf(stderr,"width:  %f\n\n",width);
				}
			}
		}


	/* now read until stop call is made */
	iplotting = 1;
	while (iplotting)
		{
		fgets(line,sizeof(line),stdin);
		if (strncmp(line,"plot",4) == 0)
			{
			sscanf(line, "plot %f %f %d",&x,&y,&ipen);
			if (x > -360.1 && x < 360.1 
				&& y > -90.1 && y< 90.1)
				{
				xymerc_(&x,&y);
				domtr2(mtx,&x,&y);
				if (ipen == IUP)
					move2(x,y);
				else if (ipen == IDN)
					line2(x,y);
				}
			}
		else if (strncmp(line,"newp",4) == 0)
			{
			sscanf(line, "newp %d",&ipen);
			setpen(ipen+1);
			}
		else if (strncmp(line,"flus",4) == 0)
			{
			for (i=0;i<210;i++)
				move2(x,y);
			makcur();
			fflush(stdout);
			fflush(stderr);
			}
		else if (strncmp(line,"stop",4) == 0 || status == EOF)
			{
			iplotting = 0;
			move2(homex,homey);
			makcur();
			mfoff(fd,0);
			termg();
			fflush(stdout);
			fflush(stderr);
			close(fd);
			}

		}

	/* all done */
	exit(0);
}
